MoltenVK/MoltenVK/GPUObjects/MVKPipeline.mm

```diff
@@ -509,40 +494,37 @@
 	if ( !_hasValidMTLPipelineStates ) { return; }
 
-	// Track dynamic state
-	const VkPipelineDynamicStateCreateInfo* pDS = pCreateInfo->pDynamicState;
-	if (pDS) {
-		for (uint32_t i = 0; i < pDS->dynamicStateCount; i++) {
-			_dynamicState.push_back(pDS->pDynamicStates[i]);
-		}
-	}
-
 	// Blending - must ignore allowed bad pColorBlendState pointer if rasterization disabled or no color attachments
 	if (_isRasterizingColor && pCreateInfo->pColorBlendState) {
-		memcpy(&_blendConstants, &pCreateInfo->pColorBlendState->blendConstants, sizeof(_blendConstants));
+		mvkCopy(_blendConstants, pCreateInfo->pColorBlendState->blendConstants, 4);
+
+		// Metal does not support blending with logic operations.
+		if (pCreateInfo->pColorBlendState->logicOpEnable && pCreateInfo->pColorBlendState->logicOp != VK_LOGIC_OP_COPY) {
+			setConfigurationResult(reportError(VK_ERROR_FEATURE_NOT_PRESENT, "Metal does not support blending using logic operations."));
+		}
+	} else {
+		static float defaultBlendConstants[4] = { 0, 0.0, 0.0, 1.0 };
+		mvkCopy(_blendConstants, defaultBlendConstants, 4);
 	}
 
 	// Topology
 	_vkPrimitiveTopology = (pCreateInfo->pInputAssemblyState && !isRenderingPoints(pCreateInfo)
 				   ? pCreateInfo->pInputAssemblyState->topology
 				   : VK_PRIMITIVE_TOPOLOGY_POINT_LIST);
 
+		// In Metal, primitive restart cannot be disabled.
+		// Just issue warning here, as it is very likely the app is not actually expecting
+		// to use primitive restart at all, and is just setting this as a "just-in-case",
+		// and forcing an error here would be unexpected to the app (including CTS).
+	if (pCreateInfo->pInputAssemblyState && !pCreateInfo->pInputAssemblyState->primitiveRestartEnable) {
+		reportWarning(VK_ERROR_FEATURE_NOT_PRESENT, "vkCreateGraphicsPipeline(): Metal does not support disabling primitive restart.");
+	}
+
 	// Rasterization
-	_mtlCullMode = MTLCullModeNone;
-	_mtlFrontWinding = MTLWindingCounterClockwise;
-	_mtlFillMode = MTLTriangleFillModeFill;
-	_mtlDepthClipMode = MTLDepthClipModeClip;
-	bool hasRasterInfo = mvkSetOrClear(&_rasterInfo, pCreateInfo->pRasterizationState);
-	if (hasRasterInfo) {
-		_mtlCullMode = mvkMTLCullModeFromVkCullModeFlags(_rasterInfo.cullMode);
-		_mtlFrontWinding = mvkMTLWindingFromVkFrontFace(_rasterInfo.frontFace);
-		_mtlFillMode = mvkMTLTriangleFillModeFromVkPolygonMode(_rasterInfo.polygonMode);
-		if (_rasterInfo.depthClampEnable) {
-			if (_device->_enabledFeatures.depthClamp) {
-				_mtlDepthClipMode = MTLDepthClipModeClamp;
-			} else {
-				setConfigurationResult(reportError(VK_ERROR_FEATURE_NOT_PRESENT, "This device does not support depth clamping."));
-			}
+	_hasRasterInfo = mvkSetOrClear(&_rasterInfo, pCreateInfo->pRasterizationState);
+	if (_hasRasterInfo) {
+		if (_rasterInfo.depthClampEnable && !_device->_enabledFeatures.depthClamp) {
+			setConfigurationResult(reportError(VK_ERROR_FEATURE_NOT_PRESENT, "This device does not support depth clamping."));
```

Comment:
Again, this should be caught by the validation layer. I know this check was already present, but since you're changing it regardless...

Reply:
Removed.